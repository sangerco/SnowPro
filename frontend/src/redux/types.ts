export interface UserData {
    username: string;
    firstName: string;
    lastName: string;
    email: string;
    avatar: string;
    bio: string;
    videos: string[];
    photos: string[];
    favMountains: string[];
  }


export const FETCH_USER_DATA_REQUEST = 'FETCH_USER_DATA_REQUEST';
export const FETCH_USER_DATA_SUCCESS = 'FETCH_USER_DATA_SUCCESS';
export const FETCH_USER_DATA_FAILURE = 'FETCH_USER_DATA_FAILURE';


export interface FetchUserDataRequestAction {
    type: typeof FETCH_USER_DATA_REQUEST
}

export interface FetchUserDataSuccessAction {
    type: typeof FETCH_USER_DATA_SUCCESS;
    payload: UserData;
}

export interface FetchUserDataFailureAction {
    type: typeof FETCH_USER_DATA_FAILURE;
    payload: string;
}

export type UserActionTypes = 
    FetchUserDataRequestAction
    | FetchUserDataSuccessAction
    | FetchUserDataFailureAction;