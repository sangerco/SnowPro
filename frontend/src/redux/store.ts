import { Action, compose } from "redux";
import { ThunkAction, configureStore } from "@reduxjs/toolkit";
import { logger } from "redux-logger";
import rootReducer from "./reducers/rootReducer";

const composeEnhancers =
  (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;

export const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(logger, composeEnhancers),
});
