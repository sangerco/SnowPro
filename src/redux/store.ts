import { configureStore, combineReducers } from "@reduxjs/toolkit";
import userReducer from "./slices/userSlice";
import authReducer from "./slices/authSlice";
import messageReducer from "./slices/messageSlice";
import messageReplyReducer from "./slices/messageReplySlice";
import reviewReducer from "./slices/reviewSlice";
import reviewReplyReducer from "./slices/reviewReplySlice";
import tagReducer from "./slices/tagSlice";
import storage from "redux-persist/lib/storage";
import { persistReducer, persistStore } from "redux-persist";

const persistConfig = {
  key: "root",
  storage,
  blacklist: ["messages", "messageReplies", "reviews", "reviewReplies"],
};

const rootReducer = combineReducers({
  users: userReducer,
  auth: authReducer,
  messages: messageReducer,
  messageReplies: messageReplyReducer,
  reviews: reviewReducer,
  reviewReplies: reviewReplyReducer,
  tags: tagReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const persistor = persistStore(store);
