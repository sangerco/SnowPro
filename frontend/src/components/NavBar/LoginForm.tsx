import React, { useState } from 'react';
import { connect } from 'react-redux';
import { sendLoginData } from '../../redux/actions/userActions';
import { RootState } from '../../redux/store';
import { Button, Input, Form, Message } from 'semantic-ui-react'

interface LoginData {
    username: string,
    password: string
};

interface LoginProps {
    user: LoginData | null;
    loading: boolean;
    error: string;
    sendLoginData: (
      username: string, 
      password: string) => Promise<void>
};

const LoginForm: React.FC<LoginProps> = ({ user, loading, error, sendLoginData }) => {
    const [ formData, setFormData ] = useState<LoginData>({
        username: '',
        password: ''
    });

    const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();

        await sendLoginData(formData.username, formData.password);

        setFormData({ username: '', password: ''});
    };

    const handleChange: React.ChangeEventHandler<HTMLInputElement> = (e) => {
        const { name, value } = e.target;
        setFormData({
            ...formData,
            [name]: value
        });
    };

    return (
        <Form onSubmit={handleSubmit} error>
            <Form.Field>
                <label>username</label>
                <input placeholder='username' name='username' value={formData.username} onChange={handleChange}/>
            </Form.Field>
            <Form.Field>
                <label>Choose Password</label>
                <Input type='password' name='password' value={formData.password} onChange={handleChange}/>
            </Form.Field>
            <Button type='submit'>Login</Button>
            {loading && <p>Loading...</p>}
            <Message
                error
                header={error}
                content={error}
            />
        </Form>
    );
}

const mapStateToProps = (state: RootState) => ({
    newUser: state.user.data,
    loading: state.user.loading,
    error: state.user.error
})
  
  export default connect(mapStateToProps, { sendLoginData })(LoginForm);