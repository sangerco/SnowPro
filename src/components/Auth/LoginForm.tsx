import React, { useState, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { AppDispatch, RootState } from "../../redux/store";
import { loginUser } from "../../redux/slices/authSlice";
import {
  Button,
  Form,
  Card,
  Dimmer,
  Loader,
  Header,
  Grid,
  Segment,
  Rating,
  Divider,
  Modal,
} from "semantic-ui-react";
import { Link, useNavigate } from "react-router-dom";
import { initialLoginState } from "../../helpers/helperStates";
import { fetchSkiAreas } from "../../redux/slices/skiAreaSlice";
import { fetchAllReviews } from "../../redux/slices/reviewSlice";

interface LoginFormProps {
  onSubmit: () => void;
}

const LoginForm: React.FC<LoginFormProps> = ({ onSubmit }) => {
  const dispatch = useDispatch<AppDispatch>();
  const navigate = useNavigate();
  const [formData, setFormData] = useState(initialLoginState);
  const [modalOpen, setModalOpen] = useState(false);
  const auth = useSelector((state: RootState) => state.auth);
  const isAuthenticated = auth.isAuthenticated;

  useEffect(() => {
    dispatch(fetchSkiAreas());
  }, [dispatch]);

  const skiAreaState = useSelector((state: RootState) => state.skiAreas);
  const skiAreas = skiAreaState.skiAreas;

  useEffect(() => {
    dispatch(fetchAllReviews());
  }, [dispatch]);

  const reviewState = useSelector((state: RootState) => state.reviews);
  const reviews = reviewState.reviews;

  if (reviews) {
    console.log(reviews[0].createdAt);
  }

  const formatDate = (date: Date) => {
    const newDate = new Date(date);
    const options: Intl.DateTimeFormatOptions = {
      year: "numeric",
      month: "long",
      day: "numeric",
    };
    return newDate.toLocaleDateString("en-US", options);
  };

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();

    dispatch(loginUser(formData));

    setFormData(initialLoginState);
  };
  useEffect(() => {
    if (isAuthenticated) {
      navigate("/");
    }
  }, [isAuthenticated, navigate]);

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { name, value } = e.target;

    setFormData((prevData) => ({ ...prevData, [name]: value }));
  };

  if (reviewState.loading || skiAreaState.loading) {
    <Dimmer active>
      <Loader>Loading...</Loader>
    </Dimmer>;
  }

  if (reviewState.error || skiAreaState.error) {
    <Dimmer active>
      <Header as="h1">Error retrieving page!</Header>
      <Button onClick={() => navigate(-1)}>Go Back</Button>
    </Dimmer>;
  }

  return (
    <>
      <Grid>
        <Grid.Row>
          <Grid.Column data-testid="skiAreaColumn" width={4}>
            {skiAreaState.loading ? (
              <Dimmer active>
                <Loader>Loading...</Loader>
              </Dimmer>
            ) : skiAreaState.error ? (
              <Dimmer active>
                <Header as="h1">
                  Error! Ski Area Data cannot be retrieved! {skiAreaState.error}
                </Header>
              </Dimmer>
            ) : skiAreas && skiAreas.length > 0 ? (
              <Card style={{ marginTop: "10px", marginLeft: "20px" }}>
                {skiAreas.map((sa) => (
                  <Card.Content key={sa.slug}>
                    <Link to={`/ski-areas/${sa.slug}`}>{sa.name}</Link>
                  </Card.Content>
                ))}
              </Card>
            ) : null}
          </Grid.Column>
          <Grid.Column width={8}>
            <Segment raised>
              <Form onSubmit={handleSubmit}>
                <Form.Field data-testid="username-input">
                  <label>
                    Username:
                    <input
                      type="text"
                      name="username"
                      value={formData.username}
                      onChange={handleChange}
                      placeholder="Username"
                    />
                  </label>
                </Form.Field>
                <Form.Field data-testid="password-input">
                  <label>
                    Password:
                    <input
                      type="password"
                      name="password"
                      value={formData.password}
                      onChange={handleChange}
                      placeholder="Password"
                    />
                  </label>
                </Form.Field>
                <Button color="blue" type="submit">
                  Login
                </Button>
              </Form>
            </Segment>
          </Grid.Column>
          <Grid.Column width={4}>
            {reviews && reviews.length > 0 ? (
              <>
                {reviews.map((review) => (
                  <Card>
                    <Card.Content key={review.id} id="review-card">
                      <Card.Header>
                        <Link to={`/ski-areas/reviews/${review.id}`}>
                          {review.header}
                        </Link>{" "}
                        <Divider />
                        {review.skiAreaName}
                      </Card.Header>
                      <Card.Description>By {review.username}</Card.Description>
                      <Card.Meta>
                        <Rating
                          icon="star"
                          defaultRating={review.stars}
                          maxRating={5}
                          disabled
                        />
                      </Card.Meta>
                      {formatDate(review.createdAt)}
                    </Card.Content>
                  </Card>
                ))}
              </>
            ) : null}
          </Grid.Column>
        </Grid.Row>
      </Grid>
      <Modal open={modalOpen}>
        <Modal.Content>
          <Header as="h1">Login Failed!</Header>
        </Modal.Content>
        <Modal.Actions>
          <Button negative onClick={() => setModalOpen(false)}>
            Close
          </Button>
        </Modal.Actions>
      </Modal>
    </>
  );
};

export default LoginForm;
