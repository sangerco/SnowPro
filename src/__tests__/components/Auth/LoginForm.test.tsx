/* eslint-disable testing-library/no-node-access */
import React from "react";
import {
  render,
  fireEvent,
  screen,
  findByTestId,
  waitFor,
} from "@testing-library/react";
import nock from "nock";
import LoginForm from "../../../components/Auth/LoginForm";
import { Provider } from "react-redux";
import { store } from "../../../redux/store";
import { BrowserRouter } from "react-router-dom";

describe("test LoginForm", () => {
  it("renders without crashing", () => {
    <Provider store={store}>
      <LoginForm />
    </Provider>;
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <LoginForm />
        </BrowserRouter>
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should contain a form", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <LoginForm />
        </BrowserRouter>
      </Provider>
    );
    const username = screen.getByTestId("username-input");
    const password = screen.getByTestId("password-input");
    expect(username).toHaveTextContent("Username:");
    expect(password).toHaveTextContent("Password:");
  });

  it("should submit the form", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .post("/login")
      .reply(200, {
        token: "faketoken",
        user: {
          username: "testuser",
          password: "password",
          firstName: "Test",
          lastName: "User",
          email: "test@user.com",
        },
      });
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get("/ski-areas")
      .reply(200, []);
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get("/ski-areas/reviews")
      .reply(200, []);
    render(
      <Provider store={store}>
        <BrowserRouter>
          <LoginForm />
        </BrowserRouter>
      </Provider>
    );

    expect(nock.pendingMocks().length).toBe(3);

    const username = screen.getByPlaceholderText("Username");
    const password = screen.getByPlaceholderText("Password");

    fireEvent.change(username, { target: { value: "testuser" } });
    fireEvent.change(password, { target: { value: "testpassword" } });

    const loginForm = await screen.findByTestId("login-form");
    fireEvent.submit(loginForm);

    await waitFor(() => {
      expect(nock.isDone()).toBe(true);
    });
  });
});
