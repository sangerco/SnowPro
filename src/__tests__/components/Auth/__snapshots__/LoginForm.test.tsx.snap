// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test LoginForm matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui grid"
  >
    <div
      class="row"
    >
      <div
        class="four wide column"
        data-testid="skiAreaColumn"
      >
        <div
          class="ui active transition visible dimmer"
          style="display: flex;"
        >
          <div
            class="content"
          >
            <div
              class="ui text loader"
            >
              Loading...
            </div>
          </div>
        </div>
      </div>
      <div
        class="eight wide column"
      >
        <div
          class="ui raised segment"
        >
          <form
            class="ui form"
            data-testid="login-form"
          >
            <div
              class="field"
              data-testid="username-input"
            >
              <label>
                Username:
                <input
                  name="username"
                  placeholder="Username"
                  type="text"
                  value=""
                />
              </label>
            </div>
            <div
              class="field"
              data-testid="password-input"
            >
              <label>
                Password:
                <input
                  name="password"
                  placeholder="Password"
                  type="password"
                  value=""
                />
              </label>
            </div>
            <button
              class="ui blue button"
              type="submit"
            >
              Login
            </button>
          </form>
        </div>
      </div>
      <div
        class="four wide column"
      />
    </div>
  </div>
</DocumentFragment>
`;
